mllib/src/main/scala/org/apache/spark/ml/tree/impl/GradientBoostedTrees.scala

```diff
@@ -205,31 +205,31 @@ private[spark] object GradientBoostedTrees extends Logging {
       case _ => data
     }
 
-    val numIterations = trees.length
-    val evaluationArray = Array.fill(numIterations)(0.0)
-    val localTreeWeights = treeWeights
-
-    var predictionAndError = computeInitialPredictionAndError(
-      remappedData, localTreeWeights(0), trees(0), loss)
-
-    evaluationArray(0) = predictionAndError.values.mean()
-
     val broadcastTrees = sc.broadcast(trees)
-    (1 until numIterations).foreach { nTree =>
-      predictionAndError = remappedData.zip(predictionAndError).mapPartitions { iter =>
-        val currentTree = broadcastTrees.value(nTree)
-        val currentTreeWeight = localTreeWeights(nTree)
-        iter.map { case (point, (pred, error)) =>
-          val newPred = updatePrediction(point.features, pred, currentTree, currentTreeWeight)
-          val newError = loss.computeError(newPred, point.label)
-          (newPred, newError)
-        }
+    val localTreeWeights = treeWeights
+    val treesIndices = trees.indices
+
+    val dataCount = remappedData.count()
+    val evaluation = remappedData
+      .map { (point: LabeledPoint) =>
+        treesIndices
+          .map(idx => {
+              val prediction = broadcastTrees.value(idx)
+                .rootNode
+                .predictImpl(point.features)
+                .prediction
+              prediction * localTreeWeights(idx)
```

Comment:
It might be very slightly more consistent to try to also use `updatePrediction` here with a 0 current prediction, just to make it pro forma consistent with the rest.